pkg/forwarders/tlsforwarder_test.go

```diff
@@ -75,126 +131,72 @@ f3cb9gYaLWdmvkx8p3g=
-			// event though 127.0.0.1 and localhost were added to alt_names in openssl config:
-			// failed to connect to reporting server: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
+			Certificates:       []tls.Certificate{tlsCert},
 			InsecureSkipVerify: true, //nolint:gosec
 		},
 	)
 	require.NoError(t, err)
-	require.NotNil(t, listener)
 	t.Logf("TLS server address %s", listener.Addr().String())
 
-	var wg sync.WaitGroup
-	go acceptLoop(t, listener, &wg,
-		[]string{
-			"<14>signal=test-signal;key=value;\n",
-			"<14>signal=test-signal-2;key=value;\n",
-		},
-	)
-
-	serializer := serializers.NewSemicolonDelimited()
-	require.NotNil(t, serializer)
-
-	log := logrusr.New(logrus.New())
-	tf, err := NewTLSForwarder(listener.Addr().String(), log,
-		func(c *tls.Config) {
-			c.RootCAs = roots
-			c.Certificates = []tls.Certificate{tlsCert}
-			c.InsecureSkipVerify = true
-		},
-	)
-	require.NoError(t, err)
-	require.NotNil(t, tf)
-
-	consumer := telemetry.NewConsumer(serializer, tf)
-
-	m, err := telemetry.NewManager(
-		"test-ping",
-		telemetry.OptManagerPeriod(time.Hour),
-		telemetry.OptManagerLogger(log),
-	)
-	require.NoError(t, err)
-
-	w := telemetry.NewWorkflow("test1")
-	p, err := provider.NewFixedValueProvider("test1-provider", types.ProviderReport{
-		"key": "value",
-	})
-	require.NoError(t, err)
-	w.AddProvider(p)
-	m.AddWorkflow(w)
-	require.NoError(t, m.AddConsumer(consumer))
-	require.NoError(t, m.Start())
+	return telemetryServer{
+		listener: listener,
+		done:     make(chan struct{}),
+	}
+}
 
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal"))
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal-2"))
-	wg.Wait()
+func (ts telemetryServer) Addr() string {
+	return ts.listener.Addr().String()
+}
 
-	require.NoError(t, listener.Close())
-	m.Stop()
+func (ts telemetryServer) Done() {
+	<-ts.done
 }
 
-func acceptLoop(t *testing.T, l net.Listener, wg *sync.WaitGroup, expectedData []string) {
+func (ts telemetryServer) RunAndAssertExpectedData(t *testing.T, expectedData []string) {
 	t.Log("server: accepting...")
-
-	// Accept just once because TLSForwarder persists the connection
-	conn, err := l.Accept()
-	if err != nil {
-		t.Logf("server: Accept() returned error: %v", err)
-		return
+	receivedData := make(chan string, len(expectedData))
+	go func() {
+		for {
+			conn, err := ts.listener.Accept()
+			if err != nil {
+				t.Logf("server: Accept() returned error: %v", err)
+				return
+			}
+			go handleConnection(t, conn, receivedData)
+		}
+	}()
+	defer ts.listener.Close()
+	for _, expected := range expectedData {
+		select {
+		case received := <-receivedData:
+			assert.Equal(t, expected, received)
+		case <-time.After(5 * time.Second):
+			assert.Failf(t, "timeout waiting for data", "expected data: %q", expected)
+		}
 	}
-	handleClient(t, conn, expectedData, wg)
+	ts.done <- struct{}{}
```

Comment:
How about we use a `sync.WaitGroup` for this?

Reply:
Do you mean having `wg sync.WaitGroup` embedded in server and at the top of this method do `wg.Add(1)` and at the bottom do `wg.Done()` and in the method `Done()` do `wg.Wait()`? I can do it like that, but it's more typing. On the other hand I may adjust the method `Done` to be like `Done() <-chan struct{}` which will make it consistent with `ctx.Done()` api...

Reply:
The reason I proposed it is because it will have more meaning: WaitGroup does what it says on the tin: you can signal if you need to wait and whether you're done. Channels can also do that but are not as expressive.

Reply:
I think channel `done`  is a pretty common pattern in Go, instead of sending I can close it here. For me, it's a little artificial to have only one call of the `wg.Add(1)` as expected. And with `WaitGroup` as a reader of code, I have to check 4 places to have the whole picture instead of 3. It felt for me more suitable previously when each test case was an increment of a `WaitGroup`

Reply:
With the proposed API we have a new method on the server: `Done()`. Which ... might signal the server that we're done and it should close. It might check if the server is done (what would be the conditions of that?). It might return a channel that will receive something on it when the server will be done (again, on what conditions? under what circumstances would that happen?). 

It's not a huge deal here, but the API is not expressive. Whereas with the sync group we could do away with the `Done()` method, pass the WaitGroup through for instance, `RunAndAssertExpectedData()` argument, and signal if we want to `Wait()` or whether we're `Done()` with it.

Reply:
I've changed it, from my perspective it may be either way. Both have pros and cons